File: assets_rebalance_frontend/src/domain/fin-assets-panel.ts
import { Result } from 'typescript-result';
import { FinAssetsGroup } from './fin-assets-group';
import { Entity } from './entity';

export class FinAssetsPanel extends Entity {
  children: FinAssetsGroup[] = []; // Initialized to an empty array
  name: string; // Required
  amountToInvest: number = 0; // Required with default value

  constructor(name: string, amountToInvest: number, children: FinAssetsGroup[] = []) {
    super()

    this.name = name;
    this.amountToInvest = amountToInvest;
    this.children = children;
  }

  get investedAmount(): number {
    return this.children.reduce((sum, group) => sum + group.currentAmount, 0);
  }

  get totalAmount(): number {
    return this.investedAmount + this.amountToInvest;
  }

  get totalScore(): number {
    return this.children.reduce((sum, group) => sum + group.score, 0);
  }

  validate(): Result<void, string[]> {
    const errors: string[] = [];

    if (this.totalScore !== 100) {
      errors.push("Panel children must sum to 100.");
    }

    for (const child of this.children) {
      const childValidation = child.validate();
      if (childValidation.isError()) {
        errors.push(...childValidation.error);
      }
    }

    return errors.length > 0 ? Result.error(errors) : Result.ok(undefined);
  }

  allAccountIds(): string[] {
    return this.children.flatMap(group =>
      group.children.map(asset => asset.accountId)
    );
  }
}

File: assets_rebalance_backend/src/Domain/FinAssetsPanel.cs
using FluentResults;
using MongoDB.Bson.Serialization.Attributes;

namespace assets_rebalance_backend.Domain;

public class FinAssetsPanel : Entity
{
    [BsonIgnore]
    public IEnumerable<FinAssetsGroup> Children { get; set; } = [];
    public required string Name { get; set; }
    public required decimal AmountToInvest { get; set; } = 0;
    public decimal InvestedAmount => Children.Sum(x => x.CurrentAmount);
    public decimal TotalAmount => AmountToInvest + InvestedAmount;
    public decimal TotalScore => Children.Sum(x => x.Score);
    public decimal TotalBuyScore => Children.Where(x => x.IsBuy(this)).Sum(x => x.Score);
    public Result Validate()
    {
        var result = Result.Ok();
        var childrenSum = Children.Sum(x => x.Score);
        if (childrenSum != 100)
            result.WithError($"panel children must sum 100 but sums {childrenSum}");

        foreach (var child in Children)
        {
            var child_validation = child.Validate();
            if (child_validation.IsFailed)
                result.WithErrors(child_validation.Errors);
        }

        return result;
    }

    public IEnumerable<Guid> AllAccountIds()
        => Children.
                SelectMany(fin_asset_group
                    => fin_asset_group
                        .Children
                        .Select(fin_asset => fin_asset.AccountId));
}
